channeldb/invoices.go

```diff
@@ -675,6 +687,314 @@ func (d *DB) UpdateInvoice(_ context.Context, ref invpkg.InvoiceRef,
+	invoice *invpkg.Invoice
+
+	// updatedAmpHtlcs holds the set of AMP HTLCs that were added or
+	// cancelled as part of this update.
+	updatedAmpHtlcs ampHTLCsMap
+
+	// settledSetIDs holds the set IDs that are settled with this update.
+	settledSetIDs map[invpkg.SetID]struct{}
+}
+
+// NOTE: this method does nothing in the k/v implementation of InvoiceUpdater.
+func (k *kvInvoiceUpdater) AddHtlc(_ models.CircuitKey,
+	_ *invpkg.InvoiceHTLC) error {
+
+	return nil
+}
+
+// NOTE: this method does nothing in the k/v implementation of InvoiceUpdater.
+func (k *kvInvoiceUpdater) ResolveHtlc(_ models.CircuitKey, _ invpkg.HtlcState,
+	_ time.Time) error {
+
+	return nil
+}
+
+// NOTE: this method does nothing in the k/v implementation of InvoiceUpdater.
+func (k *kvInvoiceUpdater) AddAmpHtlcPreimage(_ [32]byte, _ models.CircuitKey,
+	_ lntypes.Preimage) error {
+
+	return nil
+}
+
+// NOTE: this method does nothing in the k/v implementation of InvoiceUpdater.
+func (k *kvInvoiceUpdater) UpdateInvoiceState(_ invpkg.ContractState,
+	_ *lntypes.Preimage) error {
+
+	return nil
+}
+
+// NOTE: this method does nothing in the k/v implementation of InvoiceUpdater.
+func (k *kvInvoiceUpdater) UpdateInvoiceAmtPaid(_ lnwire.MilliSatoshi) error {
+	return nil
+}
+
+// UpdateAmpState updates the state of the AMP invoice identified by the setID.
+func (k *kvInvoiceUpdater) UpdateAmpState(setID [32]byte,
+	state invpkg.InvoiceStateAMP, circuitKey models.CircuitKey) error {
+
+	if _, ok := k.updatedAmpHtlcs[setID]; !ok {
+		switch state.State {
+		case invpkg.HtlcStateAccepted:
+			// If we're just now creating the HTLCs for this set
+			// then we'll also pull in the existing HTLCs that are
+			// part of this set, so we can write them all to disk
+			// together (same value)
+			k.updatedAmpHtlcs[setID] = k.invoice.HTLCSet(
+				&setID, invpkg.HtlcStateAccepted,
+			)
+
+		case invpkg.HtlcStateCanceled:
+			// Only HTLCs in the accepted state, can be cancelled,
+			// but we also want to merge that with HTLCs that may be
+			// canceled as well since it can be cancelled one by
+			// one.
+			k.updatedAmpHtlcs[setID] = k.invoice.HTLCSet(
+				&setID, invpkg.HtlcStateAccepted,
+			)
+
+			cancelledHtlcs := k.invoice.HTLCSet(
+				&setID, invpkg.HtlcStateCanceled,
+			)
+			for htlcKey, htlc := range cancelledHtlcs {
+				k.updatedAmpHtlcs[setID][htlcKey] = htlc
+			}
+
+		case invpkg.HtlcStateSettled:
+			k.updatedAmpHtlcs[setID] = make(
+				map[models.CircuitKey]*invpkg.InvoiceHTLC,
+			)
+		}
+	}
+
+	if state.State == invpkg.HtlcStateSettled {
+		// Add the set ID to the set that was settled in this invoice
+		// update. We'll use this later to update the settle index.
+		k.settledSetIDs[setID] = struct{}{}
+	}
+
+	k.updatedAmpHtlcs[setID][circuitKey] = k.invoice.Htlcs[circuitKey]
+
+	return nil
+}
+
+// Commit finalizes the update before it is written to the database.
+func (k *kvInvoiceUpdater) Commit(updateType invpkg.UpdateType) error {
```

Comment:
I wonder if we have to do a roll back when the commit fails, because currently it seems like we will change the underlying invoice in the caller function of `Commit` ?